### ISSUE TITLE
Documentation:  np.fill_diagonal val : scalar implies scalar only?

### ISSUE THREAD

**Issue by mfenner1** [numpy/numpy#6767]
In the docs for `np.fill_diagonal` (http://docs.scipy.org/doc/numpy/reference/generated/numpy.fill_diagonal.html), we have:
```
val : scalar
Value to be written on the diagonal, its type must be compatible with that of the array a.
```
Which to my reading says that it is scalars *only*.  Turns out that the main assignment in `np.fill_diagonal` (https://github.com/numpy/numpy/blob/v1.10.1/numpy/lib/index_tricks.py#L776):
```python
    a.flat[:end:step] = val
```

will take any object that can broadcast-compatible.  So, something like "val must be shape compatible with the diagonal of a" would make sense.

If this is worth a fix, I can put a PR together for it.

**Comment by mfenner1** [comment]
Also, in looking through the code for `fill_diagonal` it seems like it would be very easy to let it take an arbitrary diagonal (at least for the 2-d case):  simply adjust the starting point of the slice to `a.flat` ... for super-diagonals, just add the diagonal number to the start.  For sub-diagonal, just add (diagonal_number) * arr.shape[1].  

Not sure what the semantics of super/sub diagonals is in more than 2D.

**Comment by ltalirz** [comment]
I also just ran across this. Could one perhaps (in analogy to [numpy.put](https://docs.scipy.org/doc/numpy/reference/generated/numpy.fill_diagonal.html)) modify the documentation to something like
```
v : array_like
  Values to place on the diagonal of `a`. If `v` is shorter than the diagonal, it will be repeated as necessary.
```
If you don't want to support passing in arrays for the diagonal (I wouldn't know why not but perhaps there is a reason), please add a check in the code to prohibit it.

Otherwise, people who don't read the documentation will simply start using it this way and perhaps end up with wrong results when the functionality eventually breaks at some point.

**Comment by eric-wieser** [comment]
> it will be repeated as necessary.

I don't believe this is either accurate or desirable

**Comment by ltalirz** [comment]
> I don't believe this is either accurate or desirable

@eric-wieser Could you please explain, in which cases val will not be repeated? 

With respect to it not being desirable, would prefer an error to be raised, if the dimensions of val were not equal to either 1 or the length of the diagonal? 
The repeating of val is, however, the documented behavior of the related [numpy.put](https://docs.scipy.org/doc/numpy/reference/generated/numpy.put.html#numpy.put).

**Comment by eric-wieser** [comment]
> Could you please explain, in which cases val will not be repeated?

Ugh, you're right - I had no idea that `flatiter` had that repetition behaviour

> would you prefer an error to be raised, if the dimensions of val were not equal to either 1 or the length of the diagonal?

Yes. I don't think that that auto-repetition of non-singleton dimensions is ever a desirable feature. I'd go as far as saying that `put` is ill-designed

**Comment by eric-wieser** [comment]
Offending source code is [here](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/iterators.c#L864-L866)

**Comment by seberg** [comment]
I agree and would be +1 for making it broadcasting logic instead (and giving a deprecation warning here). I think we should try to get rid of this repeat behaviour in general, it is rarely what one would want or expect considering most of numpy.

**Comment by eric-wieser** [comment]
Perhaps `np.put(...)` should gain a parameter `val_mode='wrap'` to specify the current behaviour (since it's not that different from `np.take(mode='wrap')` and calling `put` in a way that invokes wrapping should emit a FutureWarning encouraging the explicit request?

**Comment by seberg** [comment]
Honestly, I don't care about `put`, my personal opinion is that almost always the equivalent indexing is just as good or better ;).

**Comment by mfenner1** [comment]
The conversation has gone beyond my initial documentation related comment.  But, for what it's worth, I agree with increasing the use of broadcasting logic and only using repeat behavior when it is explicitly asked for `(mode=)`.

**Comment by rjenc29** [comment]
Hi

Regarding the current (as of 1.13.3) behaviour in the case where val is array-like, I have modelled this elsewhere (hopefully accurately) by flattening val into 1-dimension and wrapping it into a ring buffer; diagonal elements are then collected by pointing at the ‘next’ item in the ring.

I am not sure if the following wording is particularly clear (or useful) but thought I would add it to the discussion:

```
val: scalar or array-like.

If val is scalar, then this is the value to be written on the diagonal.

If val is array-like, then (effectively) val is flattened to 1-dimension and wrapped to form a ring; the ‘next’ diagonal element written to a is the ‘next’ item in the ring – equivalently, the i-th diagonal element written to a is val.flat[i % len(val.flat)]. 

The type of val must be compatible with that of the array a.
```
